scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala

```diff
@@ -0,0 +1,268 @@
+package com.twitter.scalding.typed
+
+import cascading.flow.FlowDef
+import cascading.pipe.{ Each, Pipe }
+import cascading.tuple.{ Fields, TupleEntry }
+import com.twitter.scalding.TupleConverter.{ singleConverter, tuple2Converter }
+import com.twitter.scalding.TupleSetter.{ singleSetter, tup2Setter }
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.{ CascadingBinaryComparator, OrderedSerialization, Boxed }
+import scala.collection.mutable.{ Map => MMap }
+import java.util.WeakHashMap
+
+object CascadingBackend {
+  import TypedPipe._
+
+  /**
+   * we want to cache renderings of some TypedPipe to Pipe so cascading
+   * will see them as the same. Without this, it is very easy to have
+   * a lot of recomputation. Ideally we would plan an entire graph
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
+
+  private def cacheGet[T](t: TypedPipe[T], m: Mode)(p: FlowDef => CascadingPipe[T]): CascadingPipe[T] = {
+    def add(mmc: Map[Mode, CascadingPipe[Any]]): CascadingPipe[T] = {
+      val emptyFD = new FlowDef
+      val res = p(emptyFD)
+      pipeCache.put(t, mmc + (m -> res.asInstanceOf[CascadingPipe[Any]]))
+      res
+    }
+
+    pipeCache.synchronized {
+      pipeCache.get(t) match {
+        case null => add(Map.empty)
+        case somemap if somemap.contains(m) => somemap(m).asInstanceOf[CascadingPipe[T]]
+        case missing => add(missing)
+      }
+    }
+  }
+
+  final def toPipe[U](p: TypedPipe[U], fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+
+    def kvfields = Grouped.kvFields
+    val f0 = new Fields(java.lang.Integer.valueOf(0))
+
+    def singlePipe[T](t: TypedPipe[T], force: Boolean = false): CascadingPipe[T] =
+      cacheGet(t, mode) { localFD =>
+        val pipe = t.toPipe(f0)(localFD, mode, singleSetter)
+        val p = if (force) RichPipe(pipe).forceToDisk else pipe
+        CascadingPipe[T](p, f0, localFD, mode, singleConverter)
+      }
+
+    def loop[T](t: TypedPipe[T], rest: FlatMappedFn[T, U]): Pipe = t match {
+      case cp@CascadingPipe(_, fields, localFlowDef, m, _) =>
```

Comment:
I would really rather not do that since that puts the Pipe on the subclasses of TypedPipe which we are trying to avoid.

The case match is big, but each is a pretty clear block and we have a check that we have a total match.